File: src/fvmatch/data/live_feed.py
"""Best-effort live match-state feed from ESPN's public scoreboard.

No API key required. Polls the free ESPN soccer scoreboard for a league
(default FIFA World Cup, ``fifa.world``) and parses an in-progress fixture into
a :class:`LiveMatch` snapshot — current score, elapsed minute, status, and red
cards — which maps directly onto :class:`fvmatch.model.live.LiveState`.

Like :mod:`fvmatch.data.polymarket`, this is a convenience ingestion layer: it
degrades gracefully (returns ``None`` / empty) on any network or parse error and
never raises into the analysis path. The pure parsing helpers
(:func:`parse_scoreboard_event`, :meth:`LiveMatch.oriented_to`) take plain dicts
so they are unit-testable without the network.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, replace
from typing import Any

import httpx

from fvmatch.model.live import LiveState

ESPN_SCOREBOARD = "https://site.api.espn.com/apis/site/v2/sports/soccer/{league}/scoreboard"
ESPN_SUMMARY = "https://site.api.espn.com/apis/site/v2/sports/soccer/{league}/summary"
DEFAULT_LEAGUE = "fifa.world"  # FIFA World Cup
_TIMEOUT = 15.0
_MATCH_LENGTH = 90.0


@dataclass(frozen=True)
class LiveMatch:
    """Parsed live snapshot of an ESPN soccer fixture (canonical orientation).

    ``home`` / ``away`` and the goal counts are oriented to the names the caller
    matched on (see :meth:`oriented_to`), so they stay consistent with the
    Polymarket market and the Elo prior regardless of ESPN's home/away flag.
    """

    event_id: str
    home: str
    away: str
    home_goals: int
    away_goals: int
    minute: float
    status_state: str  # "pre" | "in" | "post"
    status_detail: str  # e.g. "HT", "63'", "FT"
    status_name: str = ""  # e.g. "STATUS_SECOND_HALF"
    home_team_id: str = ""
    away_team_id: str = ""
    red_cards_home: int = 0
    red_cards_away: int = 0

    @property
    def is_live(self) -> bool:
        return self.status_state == "in"

    @property
    def is_final(self) -> bool:
        return self.status_state == "post"

    @property
    def is_pre(self) -> bool:
        return self.status_state == "pre"

    def to_live_state(self) -> LiveState:
        """Map this snapshot onto the in-play model's :class:`LiveState`."""
        return LiveState(
            minute=self.minute,
            home_goals=self.home_goals,
            away_goals=self.away_goals,
            red_cards_home=self.red_cards_home,
            red_cards_away=self.red_cards_away,
            match_length=_MATCH_LENGTH,
        )

    def oriented_to(self, home_name: str, away_name: str) -> LiveMatch | None:
        """Return this match re-labelled to ``home_name``/``away_name``.

        Matches each requested name against the two ESPN competitors by token
        overlap. Returns a copy oriented so ``home`` is ``home_name`` (swapping
        goals/reds/ids if ESPN listed the teams the other way round), or ``None``
        if this event is not the requested fixture.
        """
        want_home, want_away = _name_tokens(home_name), _name_tokens(away_name)
        have_home, have_away = _name_tokens(self.home), _name_tokens(self.away)
        if _tokens_match(want_home, have_home) and _tokens_match(want_away, have_away):
            return replace(self, home=home_name, away=away_name)
        if _tokens_match(want_home, have_away) and _tokens_match(want_away, have_home):
            return replace(
                self,
                home=home_name,
                away=away_name,
                home_goals=self.away_goals,
                away_goals=self.home_goals,
                red_cards_home=self.red_cards_away,
                red_cards_away=self.red_cards_home,
                home_team_id=self.away_team_id,
                away_team_id=self.home_team_id,
            )
        return None


def _name_tokens(name: str) -> set[str]:
    """Lowercase alphanumeric tokens of a team name, minus noise words."""
    tokens = set(re.findall(r"[a-z0-9]+", name.lower()))
    return tokens - {"vs", "fc", "afc", "national", "team"}


def _tokens_match(a: set[str], b: set[str]) -> bool:
    """True if either token set is a (non-empty) subset of the other."""
    return bool(a) and bool(b) and (a <= b or b <= a)


def _coerce_int(value: Any) -> int:
    try:
        return int(str(value).strip())
    except (TypeError, ValueError):
        return 0


def _team_name(competitor: dict[str, Any]) -> str:
    team = competitor.get("team") or {}
    return str(team.get("displayName") or team.get("name") or team.get("shortDisplayName") or "")


def _minute_from_status(status: dict[str, Any], state: str) -> float:
    """Elapsed match minutes from an ESPN status block, clamped to [0, 90]."""
    if state == "pre":
        return 0.0
    if state == "post":
        return _MATCH_LENGTH
    clock = status.get("clock")
    if clock is not None:
        try:
            return min(max(float(clock) / 60.0, 0.0), _MATCH_LENGTH)
        except (TypeError, ValueError):
            pass
    # Fallback: leading digits of a display string like "63'" or "45'+2'".
    detail = str((status.get("type") or {}).get("detail") or status.get("displayClock") or "")
    m = re.match(r"\s*(\d+)", detail)
    if m:
        return min(max(float(m.group(1)), 0.0), _MATCH_LENGTH)
    return 0.0


def parse_scoreboard_event(event: dict[str, Any]) -> LiveMatch | None:
    """Parse one ESPN scoreboard ``event`` dict into a :class:`LiveMatch`.

    Oriented by ESPN's own home/away flag (use :meth:`LiveMatch.oriented_to` to
    re-label to canonical names). Returns ``None`` if the event lacks two
    competitors. Red-card counts default to 0 (filled in by network lookups).
    """
    comps = event.get("competitions") or []
    if not comps:
        return None
    comp = comps[0]
    status = comp.get("status") or event.get("status") or {}
    stype = status.get("type") or {}
    state = str(stype.get("state") or "")
    detail = str(stype.get("detail") or stype.get("shortDetail") or "")
    name = str(stype.get("name") or "")
    minute = _minute_from_status(status, state)

    competitors = comp.get("competitors") or []
    home_c = next((c for c in competitors if c.get("homeAway") == "home"), None)
    away_c = next((c for c in competitors if c.get("homeAway") == "away"), None)
    if (home_c is None or away_c is None) and len(competitors) == 2:
        home_c, away_c = competitors[0], competitors[1]
    if home_c is None or away_c is None:
        return None

    return LiveMatch(
        event_id=str(event.get("id") or ""),
        home=_team_name(home_c),
        away=_team_name(away_c),
        home_goals=_coerce_int(home_c.get("score")),
        away_goals=_coerce_int(away_c.get("score")),
        minute=minute,
        status_state=state,
        status_detail=detail,
        status_name=name,
        home_team_id=str((home_c.get("team") or {}).get("id") or ""),
        away_team_id=str((away_c.get("team") or {}).get("id") or ""),
    )


def _fetch_events(league: str) -> list[dict[str, Any]]:
    try:
        resp = httpx.get(ESPN_SCOREBOARD.format(league=league), timeout=_TIMEOUT)
        resp.raise_for_status()
        data = resp.json()
    except (httpx.HTTPError, ValueError):
        return []
    events = data.get("events") if isinstance(data, dict) else None
    return events if isinstance(events, list) else []


def _fetch_red_cards(
    league: str, event_id: str, home_team_id: str, away_team_id: str
) -> tuple[int, int]:
    """Count red cards per side from the ESPN summary ``keyEvents`` (best-effort)."""
    if not event_id:
        return (0, 0)
    try:
        resp = httpx.get(
            ESPN_SUMMARY.format(league=league),
            params={"event": event_id},
            timeout=_TIMEOUT,
        )
        resp.raise_for_status()
        data = resp.json()
    except (httpx.HTTPError, ValueError):
        return (0, 0)
    key_events = data.get("keyEvents") if isinstance(data, dict) else None
    if not isinstance(key_events, list):
        return (0, 0)
    home = away = 0
    for ke in key_events:
        if not isinstance(ke, dict):
            continue
        text = str((ke.get("type") or {}).get("text") or "").lower()
        if "red card" not in text:
            continue
        tid = str((ke.get("team") or {}).get("id") or "")
        if tid and tid == home_team_id:
            home += 1
        elif tid and tid == away_team_id:
            away += 1
    return (home, away)


def find_live_match(
    home_name: str,
    away_name: str,
    league: str = DEFAULT_LEAGUE,
    *,
    fetch_cards: bool = True,
) -> LiveMatch | None:
    """Find the ESPN fixture for ``home_name`` vs ``away_name`` in ``league``.

    Returns a :class:`LiveMatch` oriented to the requested names, with red cards
    filled in from the summary endpoint when the match is in play or finished.
    ``None`` if the fixture is not on the scoreboard or the feed is unreachable.
    """
    for event in _fetch_events(league):
        parsed = parse_scoreboard_event(event)
        if parsed is None:
            continue
        match = parsed.oriented_to(home_name, away_name)
        if match is None:
            continue
        if fetch_cards and match.status_state in ("in", "post"):
            rh, ra = _fetch_red_cards(
                league, match.event_id, match.home_team_id, match.away_team_id
            )
            match = replace(match, red_cards_home=rh, red_cards_away=ra)
        return match
    return None

File: web/src/core/liveFeed.ts
/**
 * Live match-state feed from ESPN's public scoreboard (TS port of live_feed.py).
 *
 * No API key required. Parses an in-progress fixture into a {@link LiveMatch}
 * snapshot — score, elapsed minute, status, red cards — that maps onto the
 * in-play model's {@link LiveState}. Best-effort: returns null/[] on any
 * network or parse error. The pure helpers (parseScoreboardEvent, orientedTo)
 * take plain objects so they are unit-testable without the network.
 *
 * Network fetches here must run server-side (ESPN has no CORS headers).
 */

import type { LiveState } from "./live";
import { normalizeTeamName } from "./teamNames";

const ESPN_SCOREBOARD =
  "https://site.api.espn.com/apis/site/v2/sports/soccer/{league}/scoreboard";
const ESPN_SUMMARY =
  "https://site.api.espn.com/apis/site/v2/sports/soccer/{league}/summary";
export const DEFAULT_LEAGUE = "fifa.world"; // FIFA World Cup
const MATCH_LENGTH = 90;

const NOISE_TOKENS = new Set(["vs", "fc", "afc", "national", "team"]);

export interface LiveMatch {
  eventId: string;
  home: string;
  away: string;
  homeGoals: number;
  awayGoals: number;
  minute: number;
  statusState: string; // "pre" | "in" | "post"
  statusDetail: string; // e.g. "HT", "63'", "FT"
  statusName: string; // e.g. "STATUS_SECOND_HALF"
  homeTeamId: string;
  awayTeamId: string;
  redCardsHome: number;
  redCardsAway: number;
}

// --- Loose ESPN payload shapes (only the fields we read) ---------------------

type EspnTeam = {
  id?: string | number;
  displayName?: string;
  name?: string;
  shortDisplayName?: string;
};
type EspnCompetitor = {
  homeAway?: string;
  score?: string | number;
  team?: EspnTeam;
};
type EspnStatusType = {
  state?: string;
  name?: string;
  detail?: string;
  shortDetail?: string;
};
type EspnStatus = {
  clock?: number;
  displayClock?: string;
  type?: EspnStatusType;
};
type EspnCompetition = {
  status?: EspnStatus;
  competitors?: EspnCompetitor[];
};
export type EspnEvent = {
  id?: string | number;
  name?: string;
  status?: EspnStatus;
  competitions?: EspnCompetition[];
};
type EspnScoreboard = { events?: EspnEvent[] };
type EspnKeyEvent = { type?: { text?: string }; team?: { id?: string | number } };
type EspnSummary = { keyEvents?: EspnKeyEvent[] };

// --- Pure helpers ------------------------------------------------------------

function nameTokens(name: string): Set<string> {
  const normalized = normalizeTeamName(name).toLowerCase();
  const tokens = (normalized.match(/[a-z0-9]+/g) ?? []).filter(
    (t) => !NOISE_TOKENS.has(t),
  );
  return new Set(tokens);
}

function tokensMatch(a: Set<string>, b: Set<string>): boolean {
  if (a.size === 0 || b.size === 0) return false;
  const subset = (x: Set<string>, y: Set<string>) => [...x].every((t) => y.has(t));
  return subset(a, b) || subset(b, a);
}

function coerceInt(value: unknown): number {
  const n = Number(value);
  return Number.isFinite(n) ? Math.trunc(n) : 0;
}

function teamName(competitor: EspnCompetitor): string {
  const team = competitor.team ?? {};
  return String(team.displayName ?? team.name ?? team.shortDisplayName ?? "");
}

function minuteFromStatus(status: EspnStatus, state: string): number {
  if (state === "pre") return 0;
  if (state === "post") return MATCH_LENGTH;
  const clock = status.clock;
  if (clock != null && Number.isFinite(Number(clock))) {
    return Math.min(Math.max(Number(clock) / 60, 0), MATCH_LENGTH);
  }
  // Fallback: leading digits of a display string like "63'" or "45'+2'".
  const detail = String(status.type?.detail ?? status.displayClock ?? "");
  const match = detail.match(/^\s*(\d+)/);
  if (match) return Math.min(Math.max(Number(match[1]), 0), MATCH_LENGTH);
  return 0;
}

/**
 * Parse one ESPN scoreboard event into a {@link LiveMatch}, oriented by ESPN's
 * own home/away flag. Use {@link orientedTo} to re-label to canonical names.
 */
export function parseScoreboardEvent(event: EspnEvent): LiveMatch | null {
  const comp = event.competitions?.[0];
  if (!comp) return null;

  const status = comp.status ?? event.status ?? {};
  const stype = status.type ?? {};
  const state = String(stype.state ?? "");
  const detail = String(stype.detail ?? stype.shortDetail ?? "");
  const name = String(stype.name ?? "");
  const minute = minuteFromStatus(status, state);

  const competitors = comp.competitors ?? [];
  let homeC = competitors.find((c) => c.homeAway === "home");
  let awayC = competitors.find((c) => c.homeAway === "away");
  if ((!homeC || !awayC) && competitors.length === 2) {
    [homeC, awayC] = competitors;
  }
  if (!homeC || !awayC) return null;

  return {
    eventId: String(event.id ?? ""),
    home: teamName(homeC),
    away: teamName(awayC),
    homeGoals: coerceInt(homeC.score),
    awayGoals: coerceInt(awayC.score),
    minute,
    statusState: state,
    statusDetail: detail,
    statusName: name,
    homeTeamId: String(homeC.team?.id ?? ""),
    awayTeamId: String(awayC.team?.id ?? ""),
    redCardsHome: 0,
    redCardsAway: 0,
  };
}

/**
 * Re-label a parsed match to `homeName`/`awayName` (matched by token overlap),
 * swapping goals/reds/ids if ESPN listed the teams the other way round. Returns
 * null if this event is not the requested fixture.
 */
export function orientedTo(
  match: LiveMatch,
  homeName: string,
  awayName: string,
): LiveMatch | null {
  const wantHome = nameTokens(homeName);
  const wantAway = nameTokens(awayName);
  const haveHome = nameTokens(match.home);
  const haveAway = nameTokens(match.away);

  if (tokensMatch(wantHome, haveHome) && tokensMatch(wantAway, haveAway)) {
    return { ...match, home: homeName, away: awayName };
  }
  if (tokensMatch(wantHome, haveAway) && tokensMatch(wantAway, haveHome)) {
    return {
      ...match,
      home: homeName,
      away: awayName,
      homeGoals: match.awayGoals,
      awayGoals: match.homeGoals,
      redCardsHome: match.redCardsAway,
      redCardsAway: match.redCardsHome,
      homeTeamId: match.awayTeamId,
      awayTeamId: match.homeTeamId,
    };
  }
  return null;
}

export function liveMatchToState(match: LiveMatch): LiveState {
  return {
    minute: match.minute,
    homeGoals: match.homeGoals,
    awayGoals: match.awayGoals,
    redCardsHome: match.redCardsHome,
    redCardsAway: match.redCardsAway,
    matchLength: MATCH_LENGTH,
  };
}

export function isLive(match: LiveMatch): boolean {
  return match.statusState === "in";
}

export function isFinal(match: LiveMatch): boolean {
  return match.statusState === "post";
}

// --- Network -----------------------------------------------------------------

async function fetchEvents(league: string): Promise<EspnEvent[]> {
  try {
    const res = await fetch(ESPN_SCOREBOARD.replace("{league}", league), {
      cache: "no-store",
    });
    if (!res.ok) return [];
    const data = (await res.json()) as EspnScoreboard;
    return data.events ?? [];
  } catch {
    return [];
  }
}

async function fetchRedCards(
  league: string,
  eventId: string,
  homeTeamId: string,
  awayTeamId: string,
): Promise<[number, number]> {
  if (!eventId) return [0, 0];
  try {
    const url = new URL(ESPN_SUMMARY.replace("{league}", league));
    url.searchParams.set("event", eventId);
    const res = await fetch(url.toString(), { cache: "no-store" });
    if (!res.ok) return [0, 0];
    const data = (await res.json()) as EspnSummary;
    let home = 0;
    let away = 0;
    for (const ke of data.keyEvents ?? []) {
      const text = String(ke.type?.text ?? "").toLowerCase();
      if (!text.includes("red card")) continue;
      const tid = String(ke.team?.id ?? "");
      if (tid && tid === homeTeamId) home += 1;
      else if (tid && tid === awayTeamId) away += 1;
    }
    return [home, away];
  } catch {
    return [0, 0];
  }
}

/**
 * Find the ESPN fixture for `homeName` vs `awayName` in `league`, oriented to
 * the requested names with red cards filled in when in play or finished.
 * Returns null if the fixture is not on the scoreboard or the feed is down.
 */
export async function findLiveMatch(
  homeName: string,
  awayName: string,
  league: string = DEFAULT_LEAGUE,
  options: { fetchCards?: boolean } = {},
): Promise<LiveMatch | null> {
  const { fetchCards = true } = options;
  for (const event of await fetchEvents(league)) {
    const parsed = parseScoreboardEvent(event);
    if (!parsed) continue;
    const match = orientedTo(parsed, homeName, awayName);
    if (!match) continue;
    if (fetchCards && (match.statusState === "in" || match.statusState === "post")) {
      const [redCardsHome, redCardsAway] = await fetchRedCards(
        league,
        match.eventId,
        match.homeTeamId,
        match.awayTeamId,
      );
      return { ...match, redCardsHome, redCardsAway };
    }
    return match;
  }
  return null;
}
